core/src/apps/cardano/sign_tx.py

```diff
@@ -502,27 +483,27 @@ def _cborize_shelley_witnesses(
     shelley_witnesses = []
 
     # include only one witness for each path
-    paths = set()
+    paths: List[List[int]] = list()
     for tx_input in inputs:
         if is_shelley_path(tx_input.address_n):
-            paths.add(tuple(tx_input.address_n))
+            append_unique(paths, tx_input.address_n)
     for certificate in certificates:
         if certificate.type in (
             CardanoCertificateType.STAKE_DEREGISTRATION,
             CardanoCertificateType.STAKE_DELEGATION,
         ):
-            paths.add(tuple(certificate.path))
+            append_unique(paths, certificate.path)
         elif certificate.type == CardanoCertificateType.STAKE_POOL_REGISTRATION:
             # ensured by validate_certificate:
             assert certificate.pool_parameters is not None  # validate_certificate
             for pool_owner in certificate.pool_parameters.owners:
                 if pool_owner.staking_key_path:
-                    paths.add(tuple(pool_owner.staking_key_path))
+                    append_unique(paths, pool_owner.staking_key_path)
     for withdrawal in withdrawals:
-        paths.add(tuple(withdrawal.path))
+        append_unique(paths, withdrawal.path)
 
     for path in paths:
-        witness = _cborize_shelley_witness(keychain, tx_body_hash, list(path))
+        witness = _cborize_shelley_witness(keychain, tx_body_hash, path)
```

Comment:
Isn't it better to just call `sorted(path)` here instead of introducing `append_unique` ?

```suggestion
        witness = _cborize_shelley_witness(keychain, tx_body_hash, path)
```

I am not familiar with the process, but using `sorted` has another nice deterministic property - the final list does not depend on order of elements being insterted into the set.

Reply:
that would alter the elements of the path itself if I'm not missing anything, but it's true the unique paths themselves could be sorted though I'm not sure if that could be (ab)used to infer some non-trivial relations between the keys used in the transaction (e.g. how the associated addresses are ordered within the wallet) :thinking:

Reply:
It would probably be more sensible to sort the witnesses themselves, but not sure if it's worth it given the impact on the existing test fixtures

Reply:
I'm confused by your latest comment.
Why not keep using set for key deduplication, and then calling `shelley_witnesses.sort()` in `_cborize_witnesses` to achieve determinism in the resulting list of encoded witnesses?

Reply:
Yes, it's another valid approach, the point was that altering the fixtures would result in any existing tx fixture with multiple witnesses having a 50/50 chance of being reordered (as the current approach did not rely on the resulting witnesses at all) so I'd expect it to alter quite a lot of fixtures, and also sorting, even though in place may incur some additional memory cost. I agree both are rather weak points though, sorting the witnesses is more elegant

Reply:
So I've basically reverted all the changes from https://github.com/trezor/trezor-firmware/pull/1493/commits/34d254c1911749fe63c0f22fc389b6422ef4e8b1 in https://github.com/trezor/trezor-firmware/pull/1493/commits/712774bc67fd470a1904270e809b347449ec6578 and the witnesses will be sorted. It turns out that only two testcases were impacted by the sorting and one of them is the chunking one from this PR.

Reply:
@gabrielKerekes i'm reading 5 changed `serialized_tx` fields in the respective JSON?

the changes look OK, but let's make sure we are on the same page

Reply:
Yes, two were impacted by the sorting the witnesses and the others were impacted before, when `append_unique` was introduced. The witness reordering from the `append_unique` commit caused the witnesses order to be correct also when `shelley_witnesses.sort()` is used instead.

Reply:
So 5 test cases have been impacted in total + 1 has been added.

5 test cases have been impacted by https://github.com/trezor/trezor-firmware/pull/1493/commits/34d254c1911749fe63c0f22fc389b6422ef4e8b1 when `append_unique` was used.

1 of those test cases (`Mainnet transaction with multiple inputs`) has been further impacted by the witnesses being sorted and also the chunking test case has been impacted by this.